### PR TITLE
#169817866 : Implement social authentication with Google and Facebook

### PR DESCRIPTION
**Pull request by jo-rdan**
#### What does this PR do?
It implements the login with facebook and google 
#### Description of Task to be completed?
- implement facebook authentication
- implement google authentication
#### How should this be manually tested?
- clone this repo 
- run `npm start`
**For google**
- navigate to `http://localhost:3000/api/v1/auth/google` in the browser and hit send
it should return **200** status code alongside the data and also sign the token to be able to access other routes of our app

**For facebook**
- navigate to `http://localhost:3000/api/v1/auth/facebook` in the browser and hit send
it should return **200** status code alongside the data and also sign the token to be able to access other routes of our app

#### Any background context you want to provide?
- N/A
#### What are the relevant pivotal tracker stories?
[#169817866](https://www.pivotaltracker.com/story/show/169817866)
#### Screenshots (if appropriate)

**Google**
image
image

**Facebook**
image
image

**Database**
image

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -34,6 +34,13 @@
...
       }
-    ]
+    ],
+    {
+      "parser": "babel-eslint",
+      "parserOptions": {
+        "sourceType": "module",
+        "allowImportExportEverywhere": true
+      }
+    }
```
Comment (Jajabenit250) [inline .eslintrc.json]: I think that we don't have to change eslint configuration. Check again..

**Review by Jajabenit250 — DISMISSED** [review]

File: .env.example [hunk 2]
```diff
@@ -6,3 +6,8 @@
 DATABASE_URL_DEV=VALUE
 DATABASE_URL_TEST=VALUE
 DATABASE_URL=VALUE
+GOOGLE_API=VALUE
+GOOGLE_SECRET=VALUE
+
```
Comment (Jajabenit250) [inline .env.example]: remove this space

File: .eslintrc.json [hunk 3]
```diff
@@ -35,5 +35,12 @@
         }
       }
     ]
+    // {
+    //   "parser": "babel-eslint",
+    //   "parserOptions": {
+    //     "sourceType": "module",
+    //     "allowImportExportEverywhere": true
+    //   }
+    // }
```
Comment (Jajabenit250) [inline .eslintrc.json]: Check this.. remove this comment there not necessary

File: .gitignore [hunk 4]
```diff
@@ -5,3 +5,4 @@ node_modules/
 .nyc_output/
 
 .DS_Store
+coverage/
```
Comment (Jajabenit250) [inline .gitignore]: you forgot empty  line after coverage

File: .eslintrc.json [hunk 5]
```diff
@@ -35,5 +35,12 @@
         }
       }
     ]
+    // {
```
Comment (aanchirinah) [inline .eslintrc.json]: ### Nitpick

Please remove comments when it's not being used.

File: src/controllers/socialController.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        id: profile.id,
+        firstname: profile.name.givenName,
+        lastname: profile.name.familyName,
+        email: profile.emails[0].value,
+        authtype: profile.provider
+      };
+      const {
+        email, firstname, lastname, authtype
+      } = userData;
+      users.findOrCreate({
```
Comment (aanchirinah) [inline src/controllers/socialController.js]: ### Comment

- I see `Async` but I don't see any `Await` block of code.

File: src/controllers/socialController.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+import encryptToken from '../helpers/encryption';
+import { users } from '../database/models';
+
+/**
+ * class that contains functions to implement login with social accounts
+ */
+class userController {
+  /**
+  *login functions
```
Comment (aanchirinah) [inline src/controllers/socialController.js]: ### Comment
Give better and descriptive JS Doc comments

File: src/controllers/socialController.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        }
+      });
+      done(null, userData);
+    } catch (error) {
+      done(error, false, error.message);
+    }
+  }
+
+  /**
+  *login function
```
Comment (aanchirinah) [inline src/controllers/socialController.js]: ### Nitpick
- Write a descriptive JS Doc.

File: src/database/models/users.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
+
+module.exports = (sequelize, DataTypes) => {
+  const users = sequelize.define('users', {
+    firstname: DataTypes.STRING,
```
Comment (aanchirinah) [inline src/database/models/users.js]: ### Critical
- Define this with other constraints, like primary key, unique, foreign key and data types, `NOT NULL` etc.

File: src/services/oauth.js [hunk 10]
```diff
@@ -0,0 +1,18 @@
+import passport from 'passport';
+import GooglePlusTokenStrategy from 'passport-google-plus-token';
+import FacebookToken from 'passport-facebook-token';
+import { config } from 'dotenv';
+import userController from '../controllers/socialController';
+
+config();
+passport.use('google',
```
Comment (aanchirinah) [inline src/services/oauth.js]: ### Nitpick

- This can be written better

File: src/tests/socialAuthTests.js [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+  });
+  it('should return user data', async (done) => {
+    const accessToken = 'zzzz-zzzz-zzzz';
+    const refreshToken = 'zzzz-zzzz-zzzz';
+    const callBack = sinon.spy();
+    userController.googleAndFacebookPlusAuth(accessToken, refreshToken, profile, callBack);
+    expect(callBack.withArgs({ socialUser }));
+    done();
+  });
+  it('should not return user data', async (done) => {
```
Comment (aanchirinah) [inline src/tests/socialAuthTests.js]: ### Nitpick

- Make you test spec descriptions be more descriptive.

File: src/tests/socialMock.js [hunk 12]
```diff
@@ -0,0 +1,47 @@
+const profile = {
```
Comment (aanchirinah) [inline src/tests/socialMock.js]: ### Nitpick
- I'll advise you to use `faker.js` also for mocking data.

**Review by aanchirinah — CHANGES_REQUESTED** [review]

File: src/database/models/users.js (hunk 9 above)
Comment (jo-rdan) [reply src/database/models/users.js]: for the primary key it is already defined in the migration file for users. and about the foreign keys we don't have relationships yet we will be using them later in the project

File: src/controllers/socialController.js [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        id: profile.id,
+        firstname: profile.name.givenName,
+        lastname: profile.name.familyName,
+        email: profile.emails[0].value,
+        authtype: profile.provider
+      };
+      const {
+        email, firstname, lastname, authtype
+      } = userData;
+      users.findOrCreate({
```
Comment (aanchirinah) [inline src/controllers/socialController.js]: ### Critical

- Handle this asynchronously, use known patterns like `Async/await` or `then/catch`

**Review by Jajabenit250 — APPROVED** [review]
Great Work Mate you did it. Keep the energy.

File: src/services/oauth.js [hunk 14]
```diff
@@ -0,0 +1,16 @@
+import GooglePlusTokenStrategy from 'passport-google-plus-token';
+import FacebookToken from 'passport-facebook-token';
+import { config } from 'dotenv';
+import userController from '../controllers/socialController';
+import connectMiddleware from '../helpers/passportStrategies';
+
+config();
+connectMiddleware('google', new GooglePlusTokenStrategy({
```
Comment (aanchirinah) [inline src/services/oauth.js]: ### Nipitck

- Can be done differently

**Review by aanchirinah — CHANGES_REQUESTED** [review]

File: src/controllers/socialController.js (hunk 13 above)
Comment (jo-rdan) [reply src/controllers/socialController.js]: I used async and await

**Pull request merged**